Add summary to setup action (#23)
gradle/github-actions: README.md

A collection of composite GitHub Actions

## Add a workflow summary to Maven builds 

### Description

> The `maven-setup` action adds a workflow summary for each Maven invocation. 

Reviewer: for my understanding, the workflow summary is available even for builds that do not publish scans?

Author: Yes

> The summary can also be optionally added as a pull-request comment (if in a pull-request context). 

Reviewer: If this is enabled by default, I would adjust the wording here.
```suggestion
The summary is also added as a pull-request comment in a pull-request context. This can be disabled via <link> 
```

This action can be combined with the `maven-publish-build-scan` action to publish Build Scan® in a fork context. See [the relevant section](#publish-build-scans-for-pull-requests-issued-from-forked-repositories) for details.

### Usage

Insert the `Setup Maven` step once in each job having steps invoking Maven. Make sure to put the step before the Maven invocation.

```yaml
name: PR Build
jobs:
  build:  

>       - name: Setup Maven

Reviewer: I know this is a bit late here, but I was just thinking, since this doesn't set up maven itself, perhaps the name is misleading. Does this add maven to the path?
I can't think of a better name, maybe `maven-summary-setup`? I wonder what @bigdaz would think here.

Author: We already discussed this in the [naming specification](https://docs.google.com/document/d/1CUt2IQIaBK6D7xihD2MT9_39Zvi3tRtuSW1dhnB7GuE/edit#heading=h.pr5426ll05no) and decided to keep it generic to allow further capabilities in the future (CI injection?)

        uses: gradle/github-actions/maven-setup@v1
      - name: Build with Maven
        run: ./mvnw clean package
[...]
```

### Implementation details

The action enables two features:
- Display a summary of the Maven builds as a GitHub workflow summary or as a pull-request comment
- Capture unpublished Build Scan® data as a workflow artifact per job with prefix `build-scan-data-maven`, which can then be published in a dependent workflow.

#### Summary

By default, a summary will be added to the GitHub workflow calling the action (can be skipped with `skip-job-summary` is set to `true`):

![workflow](./doc/summary-workflow.png)

> By default, the pull-request will be commented with the summary (can be skipped if `skip-pr-comment` is set to `true`):

Reviewer: ```suggestion
By default, a comment will be added to the pull-request with the summary (can be skipped if `skip-pr-comment` is set to `true`):
```

I like that we're adding this comment by default. I think it is the best user experience but it does differ from how the Gradle action works.

      "requestedTasks": "install",
      "buildId": "1701252760831-89e3583d-3c3c-4a64-a958-e61aa791f4f4",
      "buildFailure": true,
      "buildTimestamp": "1701252760800",
      "buildScanLink": "https://<DEVELOCITY_URL>/s/dxqnlj5hgybq4"
    }
  ]
}
```
#### Capture unpublished Build Scan®

The _capture strategy_ can be customized:

> - `ALWAYS`: default behavior, capture will be attempted on each Maven invocation

Reviewer: I think we discussed this before, but the next Develocity maven extension will not feature these values like `ALWAYS`, `ON_FAILURE`, and `ON_DEMAND`. It will only have conditional publication. `onlyIf`. Can you remind me why we didn't align this API?

Author: as far as I remember, I didn't want to execute arbitrary code in the Maven process. That would also force to ship a parser / interpreter.

Reviewer: got it, thanks

- `ON_FAILURE`: capture will be attempted only on failed builds
- `ON_DEMAND`: capture will be attempted if `CAPTURE_BUILD_SCAN=true` in the environment

The _capture_ can be _enabled_/_disabled_ separately:
- `capture-unpublished-build-scans`: to disable unpublished Build Scan® capture
- `capture-build-scan-links`: to disable Build Scan® link capture

> The process is handled by a [Maven extension](https://maven.apache.org/guides/mini/guide-using-extensions.html) `maven-build-scan-capture-extension.jar` which is running during each Maven invocation.

Reviewer: I wonder if this additional maven extension could cause build cache misses. Have we tested if it causes "goal implementation changed"?

Author: There is no "goal implementation changed" with the extension applied, the cache keys are consistent.
See [this comparison](https://ge.solutions-team.gradle.com/c/dmzfighnwjyro/n4smq2zqiug5k/goal-inputs) for instance

Reviewer: Very nice!

The extension is automatically registered by configuring the environment `MAVEN_OPTS=-Dmaven.ext.classpath=<PATH_TO_EXTENSION>`.

_Note:_<br>
If `MAVEN_OPTS` environment variable is set in the step invoking the `mvn` command, the extension won't be registered.
Make sure to use `MAVEN_OPTS: ${{ env.MAVEN_OPTS }} <EXTRA_PARAMETERS>` construction to append the extra parameters and have the extension registered.

The captured files are added as workflow artifact (one artifact per job).

**Event Triggers**:

- `pull_request`: To capture unpublished Build Scan®
- `workflow_run`: To capture Build Scan® links

**Permissions**:

The following permissions are required for this action to operate:
- `pull-requests: write`: to comment the pull-request

**Action inputs**:

| Name                              | Description                                                                  | Default  |
|-----------------------------------|------------------------------------------------------------------------------|----------|
| `capture-strategy`                | *Optional*: Build Scan capture strategy (ALWAYS, ON_FAILURE, ON_DEMAND)      | `ALWAYS` |

> | `capture-unpublished-build-scans` | *Optional*: Whether to enable or not unpublished Build Scans capture         | `true`   |

Reviewer: Ok so two thoughts here.
1. This (and the others) should say `Whether or not to enable unpublished scans capture`

2. Can we just get rid of all the `whether or not` and instead say `Enable Build Scan links capture` to make it more concise?

Author: Thanks, will apply suggestion 2., it is way more readable 👍

| `capture-build-scan-links`        | *Optional*: Whether to enable or not Build Scan links capture                | `true`   |
| `job-name`                        | *Optional*: Job name (used in summary and as artifact suffix)                |          |
| `add-pr-comment`                  | *Optional*: Whether to comment or not the pull-request with Build Scan links | `true`   |
| `add-job-summary`                 | *Optional*: Whether to add or not a summary to the GitHub workflow           | `true`   |
| `add-project-id-in-job-summary`   | *Optional*: Whether to add or not project id to the job summary / comment    | `false`  |  

## Publish Build Scans® for pull-requests issued from forked repositories

### Description

> When submitting a pull request, a GitHub workflow that validates the change is usually triggered, however the Develocity Build Scans® can’t be published 

Reviewer: ```suggestion
When submitting a pull request from a forked Github repository, a GitHub workflow that validates the change cannot publish a Build Scan® 
```

> as workflows from forked repositories do not have access to secrets although an access token is required to publish a Build Scan®

Reviewer: ```suggestion
as workflows from forked repositories do not have access to secrets. An access token is required to publish a Build Scan®.
```

          develocity-url: 'https://<MY_DEVELOCITY_URL>'
          develocity-access-key: ${{ secrets.<DEVELOCITY_ACCESS_KEY> }}
```

_Note:_<br>
Some parameters need to be adjusted here:
- The workflow name (here `PR Build`) triggered when a pull-request is submitted
- The Develocity URL (here `https://<MY_DEVELOCITY_URL>`)
- The secret name holding the Develocity access key (here `<DEVELOCITY_ACCESS_KEY>`)
 
#### Implementation details

The action will download any saved Build Scan® and publish them to Develocity.

> The list of pull-request authors allowed to publish a Build Scan® can be specified by the csv parameter `authorized-users-list`. 

Reviewer: What is the behavior if this list is missing? Are no scans published or are all scans published?

Author: All scans would be published, I changed the comment to:
> The action will publish Build Scans® if the initial pull-request author belongs to the list or if the list is empty.
